[UX] App no longer starts up after APM app routing changes

After https://github.com/elastic/kibana/pull/104274 was merged, the UX no longer starts up, as it uses APM app components that expect @kbn/typed-react-router-config being used, but that is not the case for the UX app. The following error appears in the console:

<details>

```
react_devtools_backend.js:2560 Error: Router not found in context
    at useRouter (use_router.js:17)
    at useParams (use_params.js:11)
    at useApmParams (use_apm_params.ts:9)
    at EnvironmentFilter (index.tsx:47)
    at us (kbn-ui-shared-deps.js:390)
    at wa (kbn-ui-shared-deps.js:390)
    at ml (kbn-ui-shared-deps.js:390)
    at fl (kbn-ui-shared-deps.js:390)
    at ol (kbn-ui-shared-deps.js:390)
    at kbn-ui-shared-deps.js:390
overrideMethod @ react_devtools_backend.js:2560
da @ kbn-ui-shared-deps.js:390
Ca.n.callback @ kbn-ui-shared-deps.js:390
yo @ kbn-ui-shared-deps.js:390
vo @ kbn-ui-shared-deps.js:390
vl @ kbn-ui-shared-deps.js:390
t.unstable_runWithPriority @ kbn-ui-shared-deps.js:398
$r @ kbn-ui-shared-deps.js:390
_l @ kbn-ui-shared-deps.js:390
ol @ kbn-ui-shared-deps.js:390
(anonymous) @ kbn-ui-shared-deps.js:390
t.unstable_runWithPriority @ kbn-ui-shared-deps.js:398
$r @ kbn-ui-shared-deps.js:390
Gr @ kbn-ui-shared-deps.js:390
z @ kbn-ui-shared-deps.js:398
C.port1.onmessage @ kbn-ui-shared-deps.js:398
15:22:31.006 kbn-ui-shared-deps.js:398 Uncaught Error: Router not found in context
    at useRouter (use_router.js:17)
    at useParams (use_params.js:11)
    at useApmParams (use_apm_params.ts:9)
    at EnvironmentFilter (index.tsx:47)
    at us (kbn-ui-shared-deps.js:390)
    at wa (kbn-ui-shared-deps.js:390)
    at ml (kbn-ui-shared-deps.js:390)
    at fl (kbn-ui-shared-deps.js:390)
    at ol (kbn-ui-shared-deps.js:390)
    at kbn-ui-shared-deps.js:390
useRouter @ use_router.js:17
useParams @ use_params.js:11
useApmParams @ use_apm_params.ts:9
EnvironmentFilter @ index.tsx:47
us @ kbn-ui-shared-deps.js:390
wa @ kbn-ui-shared-deps.js:390
ml @ kbn-ui-shared-deps.js:390
fl @ kbn-ui-shared-deps.js:390
ol @ kbn-ui-shared-deps.js:390
(anonymous) @ kbn-ui-shared-deps.js:390
t.unstable_runWithPriority @ kbn-ui-shared-deps.js:398
$r @ kbn-ui-shared-deps.js:390
Gr @ kbn-ui-shared-deps.js:390
z @ kbn-ui-shared-deps.js:398
C.port1.onmessage @ kbn-ui-shared-deps.js:398
```
</details>

--- a/packages/kbn-typed-react-router-config/src/create_router.test.tsx
+++ b/packages/kbn-typed-react-router-config/src/create_router.test.tsx
@@ -158,6 +158,12 @@ describe('createRouter', () => {
         router.getParams('/service-map', history.location);
       }).toThrowError('No matching route found for /service-map');
     });
+
+    it('does not throw an error if the given path does not match any routes but is marked as optional', () => {
+      expect(() => {
+        router.getParams('/service-map', history.location, true);
+      }).not.toThrowError();
+    });
   });
 
   describe('matchRoutes', () => {

--- a/packages/kbn-typed-react-router-config/src/create_router.ts
+++ b/packages/kbn-typed-react-router-config/src/create_router.ts
@@ -45,10 +45,12 @@ export function createRouter<TRoutes extends Route[]>(routes: TRoutes): Router<T
   const matchRoutes = (...args: any[]) => {
     let path: string = args[0];
     let location: Location = args[1];
+    let optional: boolean = args[2];
 
     if (args.length === 1) {
       location = args[0] as Location;
       path = location.pathname;
+      optional = args[1];
     }
 
     const greedy = path.endsWith('/*') || args.length === 1;
@@ -64,6 +66,9 @@ export function createRouter<TRoutes extends Route[]>(routes: TRoutes): Router<T
       : findLastIndex(matches, (match) => match.route.path === path);
 
     if (matchIndex === -1) {
+      if (optional) {
+        return [];
+      }
       throw new Error(`No matching route found for ${path}`);
     }
 
@@ -144,9 +149,11 @@ export function createRouter<TRoutes extends Route[]>(routes: TRoutes): Router<T
     link: (path, ...args) => {
       return link(path, ...args);
     },
-    getParams: (path, location) => {
-      const matches = matchRoutes(path, location);
-      return merge({ path: {}, query: {} }, ...matches.map((match) => match.match.params));
+    getParams: (...args: any[]) => {
+      const matches = matchRoutes(...args);
+      return matches.length
+        ? merge({ path: {}, query: {} }, ...matches.map((match) => match.match.params))
+        : undefined;
     },
     matchRoutes: (...args: any[]) => {
       return matchRoutes(...args) as any;

--- a/packages/kbn-typed-react-router-config/src/types/index.ts
+++ b/packages/kbn-typed-react-router-config/src/types/index.ts
@@ -125,6 +125,11 @@ export interface Router<TRoutes extends Route[]> {
     path: TPath,
     location: Location
   ): OutputOf<TRoutes, TPath>;
+  getParams<TPath extends PathsOf<TRoutes>, TOptional extends boolean>(
+    path: TPath,
+    location: Location,
+    optional: TOptional
+  ): TOptional extends true ? OutputOf<TRoutes, TPath> | undefined : OutputOf<TRoutes, TPath>;
   link<TPath extends PathsOf<TRoutes>>(
     path: TPath,
     ...args: TypeAsArgs<TypeOf<TRoutes, TPath>>

--- a/packages/kbn-typed-react-router-config/src/use_params.ts
+++ b/packages/kbn-typed-react-router-config/src/use_params.ts
@@ -9,9 +9,9 @@
 import { useLocation } from 'react-router-dom';
 import { useRouter } from './use_router';
 
-export function useParams(path: string) {
+export function useParams(path: string, optional: boolean = false) {
   const router = useRouter();
   const location = useLocation();
 
-  return router.getParams(path as never, location);
+  return router.getParams(path as never, location, optional);
 }

--- a/x-pack/plugins/apm/public/application/uxApp.tsx
+++ b/x-pack/plugins/apm/public/application/uxApp.tsx
@@ -10,7 +10,8 @@ import euiLightVars from '@elastic/eui/dist/eui_theme_light.json';
 import { AppMountParameters, CoreStart } from 'kibana/public';
 import React from 'react';
 import ReactDOM from 'react-dom';
-import { Route, Router } from 'react-router-dom';
+import { Route as ReactRouterRoute } from 'react-router-dom';
+import { RouterProvider, createRouter } from '@kbn/typed-react-router-config';
 import { DefaultTheme, ThemeProvider } from 'styled-components';
 import { i18n } from '@kbn/i18n';
 import type { ObservabilityRuleTypeRegistry } from '../../../observability/public';
@@ -66,12 +67,14 @@ function UxApp() {
       })}
     >
       <div data-test-subj="csmMainContainer" role="main">
-        <Route component={ScrollToTopOnPathChange} />
+        <ReactRouterRoute component={ScrollToTopOnPathChange} />
         <RumHome />
       </div>
     </ThemeProvider>
   );
 }
+
+const uxRouter = createRouter([]);
 
 export function UXAppRoot({
   appMountParameters,
@@ -107,12 +110,12 @@ export function UXAppRoot({
           services={{ ...core, ...plugins, embeddable, data }}
         >
           <i18nCore.Context>
-            <Router history={history}>
+            <RouterProvider history={history} router={uxRouter}>
               <UrlParamsProvider>
                 <UxApp />
                 <UXActionMenu appMountParameters={appMountParameters} />
               </UrlParamsProvider>
-            </Router>
+            </RouterProvider>
           </i18nCore.Context>
         </KibanaContextProvider>
       </ApmPluginContext.Provider>

--- a/x-pack/plugins/apm/public/components/shared/EnvironmentFilter/index.tsx
+++ b/x-pack/plugins/apm/public/components/shared/EnvironmentFilter/index.tsx
@@ -64,12 +64,15 @@ function getOptions(environments: string[]) {
 export function EnvironmentFilter() {
   const history = useHistory();
   const location = useLocation();
-  const { path } = useApmParams('/*');
+  const apmParams = useApmParams('/*', true);
   const { urlParams } = useUrlParams();
 
   const { environment, start, end } = urlParams;
   const { environments, status = 'loading' } = useEnvironmentsFetcher({
-    serviceName: 'serviceName' in path ? path.serviceName : undefined,
+    serviceName:
+      apmParams && 'serviceName' in apmParams.path
+        ? apmParams.path.serviceName
+        : undefined,
     start,
     end,
   });

--- a/x-pack/plugins/apm/public/hooks/use_apm_params.ts
+++ b/x-pack/plugins/apm/public/hooks/use_apm_params.ts
@@ -9,7 +9,17 @@ import { OutputOf, PathsOf, useParams } from '@kbn/typed-react-router-config';
 import { ApmRoutes } from '../components/routing/apm_route_config';
 
 export function useApmParams<TPath extends PathsOf<ApmRoutes>>(
+  path: TPath,
+  optional: true
+): OutputOf<ApmRoutes, TPath> | undefined;
+
+export function useApmParams<TPath extends PathsOf<ApmRoutes>>(
   path: TPath
-): OutputOf<ApmRoutes, TPath> {
-  return useParams(path as never);
+): OutputOf<ApmRoutes, TPath>;
+
+export function useApmParams(
+  path: string,
+  optional?: true
+): OutputOf<ApmRoutes, PathsOf<ApmRoutes>> | undefined {
+  return useParams(path, optional);
 }

